# introduction:

The MovieLens dataset is a rich source of information about movies, user ratings, and tags. This report aims to explore the dataset using the provided modules, identifying patterns and insights into the film industry's trends, user preferences, and the characteristics of popular movies.

By leveraging the methods implemented in ratings.py, movies.py, tags.py, and links.py, we analyze the data to uncover:

The distribution of movies by release year and genre.
Audience engagement through ratings and tags.
Notable trends in movie production, directors, and budgets.
The dataset's breadth and variety allow us to delve into numerous aspects of cinema, providing a comprehensive overview of the film world.

In [2]:
from ratings import Ratings
from movies import Movies
from links import Links
from tags import Tags


ratings_file = "ratings.csv"
movies_file = "movies.csv"
tags_file = "tags.csv"
links_file = "links.csv"
fields_file = "fields.csv"

ratings = Ratings(ratings_file, movies_file)
movies = ratings.Movies(ratings.ratings_file, ratings.movie_titles)
movies_method = Movies(movies_file)
tags = Tags(tags_file)
links = Links(links_file, fields_file)


# Rating

Im going to explore the dataset 

Distribution of movies by release year:

The analysis of movies by their release year provides insights into the production trends of the film industry

In [13]:
%timeit release_year_dist = movies.dist_by_year()

# Display the distribution
release_year_dist = movies.dist_by_year()
for year, count in release_year_dist.items():
    print(f"Year: {year}, Count: {count}")

55.9 ms ± 1.08 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Year: 1996, Count: 6040
Year: 1997, Count: 1916
Year: 1998, Count: 507
Year: 1999, Count: 2439
Year: 2000, Count: 10061
Year: 2001, Count: 3922
Year: 2002, Count: 3478
Year: 2003, Count: 4014
Year: 2004, Count: 3279
Year: 2005, Count: 5813
Year: 2006, Count: 4059
Year: 2007, Count: 7114
Year: 2008, Count: 4351
Year: 2009, Count: 4158
Year: 2010, Count: 2301
Year: 2011, Count: 1690
Year: 2012, Count: 4656
Year: 2013, Count: 1664
Year: 2014, Count: 1439
Year: 2015, Count: 6616
Year: 2016, Count: 6703
Year: 2017, Count: 8198
Year: 2018, Count: 6418


Key points:
2000 had the highest number of movie releases, it is like 10061 movies in one year
Abd the lowest number is 1998, it had only 507 movies 

Distribution of Movies by Rating

Analyzing movie ratings provides insights into audience preferences and the overall reception of movies

code explaination:

%timeit: Measures the time it takes to execute movies.dist_by_release(), providing insights into the method's efficiency.

movies.dist_by_release(): Returns a dictionary where keys are years, and values are the number of movies released in those years.

For Loop: Iterates through the dictionary to print each year and the corresponding movie count.

Purpose: This analysis helps identify periods of high and low movie production, reflecting historical or technological changes in the industry.

In [14]:
# Im analyzing the distribution by rating
%timeit rating_dist = movies.dist_by_rating()

# now im gonna display the distribution
rating_dist = movies.dist_by_rating()
for rating, count in rating_dist.items():
    print(f"Rating: {rating}, Count: {count}")


17.1 ms ± 754 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rating: 0.5, Count: 1370
Rating: 1.0, Count: 2811
Rating: 1.5, Count: 1791
Rating: 2.0, Count: 7551
Rating: 2.5, Count: 5550
Rating: 3.0, Count: 20047
Rating: 3.5, Count: 13136
Rating: 4.0, Count: 26818
Rating: 4.5, Count: 8551
Rating: 5.0, Count: 13211


Key points:
The highest count is for 4.0 stars (26,818 movies), suggesting a generally favorable audience reception.
Ratings of 0.5 and 1.0 combined account for 4,181 movies, indicating a critical segment of the audience.

Top movies by number of ratings

The top-rated movies provide a snapshot of audience engagement and popularity.

In [15]:
# Im gonna get top 5 movies by number of ratings
%timeit top_movies_by_count = movies.top_by_num_of_ratings(5)

# again im gonna display the top movies
top_movies_by_count = movies.top_by_num_of_ratings(5)
for movie, count in top_movies_by_count.items():
    print(f"Movie: {movie}, Number of Ratings: {count}")


24.6 ms ± 961 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Movie: 356,Forrest Gump (1994), Number of Ratings: 329
Movie: 318,Shawshank Redemption, The (1994), Number of Ratings: 317
Movie: 296,Pulp Fiction (1994), Number of Ratings: 307
Movie: 593,Silence of the Lambs, The (1991), Number of Ratings: 279
Movie: 2571,Matrix, The (1999), Number of Ratings: 278


Key points:
Forrest Gump leads with 329 ratings.
Four out of the five movies are from the 1990s, highlighting the enduring appeal of films from this decade.

A significant increase in movie releases during the early 21st century suggests a boom in production, likely driven by advancements in digital technology and globalization.
Specific years, such as [insert year with the highest count], stand out as peaks, potentially linked to economic or cultural shifts.

Conclusion:
This analysis of the MovieLens dataset reveals several key insights:
A surge in movie production occurred in the early 2000s, with continued high numbers in recent years.

Audience ratings are generally positive, with 4.0 stars being the most common rating.

The most rated movies are culturally significant classics, particularly from the 1990s.

This data provides a foundation for further exploration into user preferences, genre trends, and the evolution of the film industry.

# Movies:

Genres define the identity of movies and cater to diverse audience tastes. Analyzing their distribution helps identify dominant genres.

In [3]:
# Analyze the distribution by genres
%timeit genre_dist = movies_method.dist_by_genres()

# Display the distribution
genre_dist = movies_method.dist_by_genres()
print("Distribution of Movies by Genre:")
for genre, count in genre_dist.items():
    print(f"Genre: {genre}, Count: {count}")


988 μs ± 51.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Distribution of Movies by Genre:
Genre: Drama, Count: 507
Genre: Comedy, Count: 365
Genre: Romance, Count: 208
Genre: Thriller, Count: 179
Genre: Action, Count: 158
Genre: Adventure, Count: 126
Genre: Crime, Count: 122
Genre: Children, Count: 100
Genre: Fantasy, Count: 69
Genre: Sci-Fi, Count: 69
Genre: Mystery, Count: 58
Genre: Musical, Count: 53
Genre: Horror, Count: 51
Genre: War, Count: 48
Genre: Animation, Count: 37
Genre: Documentary, Count: 25
Genre: Western, Count: 23
Genre: Film-Noir, Count: 18
Genre: IMAX, Count: 3


Keypoints:
The most common genres ([list top genres]) indicate audience preferences for certain storytelling styles.
Less common genres suggest niche markets that cater to specific audience groups.

The method for analyzing the genre distribution performed efficiently:

Execution Time: 988 µs ± 51.1 µs per loop (mean ± standard deviation over 7 runs with 1,000 loops each).


Dominant Genres:

Drama leads the chart with 507 movies, reflecting its broad appeal and storytelling flexibility.
Comedy is the second-most popular genre with 365 movies, showcasing its universal entertainment value.

Moderately Represented Genres:

Romance, Thriller, Action, and Adventure each have significant representation, indicating their steady popularity with audiences.

Least Represented Genres:

IMAX has only 3 movies, as it typically refers to a special format rather than a genre.
Film-Noir and Western are niche genres, likely reflecting their historical or stylistic specificity.


Conclusion

The data reveals that Drama and Comedy are the most prevalent genres, likely due to their universal appeal. On the other hand, niche genres like Film-Noir and Western cater to a smaller, specific audience. This distribution highlights the industry's focus on genres that maximize audience engagement while still catering to diverse tastes.

code explaination:
%timeit: Measures the execution time of movies.dist_by_genres().

movies.dist_by_genres(): Returns a dictionary where keys are genres, and values are the number of movies in each genre.

For Loop: Prints each genre and its count.

Purpose: Identifying popular genres provides insights into audience preferences and market trends.



# Tags:

Tags capture audience sentiments and categorize movies based on themes or characteristics.

In [3]:
# Analyze the most popular tags
%timeit popular_tags = tags.most_popular(5)

# Display the tags
popular_tags = tags.most_popular(5)
print("Most Popular Tags:")
for tag, count in popular_tags.items():
    print(f"Tag: {tag}, Count: {count}")


1.97 ms ± 165 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Most Popular Tags:
Tag: In Netflix queue, Count: 131
Tag: atmospheric, Count: 36
Tag: superhero, Count: 24
Tag: thought-provoking, Count: 24
Tag: funny, Count: 23


keypoints:

Popular tags like "In Netfix queue, atmospheric, superhero, thought-provoking, funny" indicate recurring themes or features that resonate with audiences.
The frequency of these tags highlights their importance in defining movies and their appeal.

Code explanation:

Tags represent keywords or themes associated with movies, reflecting audience perceptions.

%timeit: Measures how efficiently the method executes.

tags.most_popular(5): Retrieves the top 5 most popular tags, returning a dictionary where keys are tags and values are their counts.

For Loop: Prints each tag and its count.

Purpose: This helps understand recurring themes or descriptors that resonate with audiences.



Objective
This section identifies the most popular tags in the dataset. Tags provide a way to categorize and describe movies, reflecting audience perceptions and preferences.

Performance
The method for analyzing popular tags performed as follows:

Execution Time: 1.97 ms ± 165 µs per loop (mean ± standard deviation over 7 runs with 100 loops each).


Dominant Tag:

In Netflix queue is by far the most popular tag with 131 mentions. This indicates a strong association of the dataset with Netflix content, highlighting its role in the modern movie-watching experience.

Recurring Themes:

Tags like atmospheric, thought-provoking, and superhero showcase audience preferences for immersive experiences, engaging content, and larger-than-life storytelling.
The tag funny underscores the enduring appeal of humor in entertainment.

Diversity in Tags:

While some tags reflect content format (In Netflix queue), others emphasize movie attributes (atmospheric, superhero), demonstrating the variety in how movies are perceived and categorized.


Conclusion

The data reveals that tags are a valuable tool for understanding audience sentiments and preferences. The prominence of In Netflix queue highlights the influence of streaming platforms in shaping viewing habits, while tags like thought-provoking and funny underline the importance of emotional and intellectual engagement in movie experiences.

# Links

This section identifies directors with the highest number of movies.

In [4]:
# Analyze the top directors
%timeit top_directors = links.top_directors(5)

# Display the top directors
top_directors = links.top_directors(5)
print("Top Directors by Number of Movies:")
for director, count in top_directors.items():
    print(f"Director: {director}, Number of Movies: {count}")


6.91 ms ± 359 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Top Directors by Number of Movies:
Director: Woody Allen, Number of Movies: 46
Director: Alfred Hitchcock, Number of Movies: 37
Director: Steven Spielberg, Number of Movies: 32
Director: Clint Eastwood, Number of Movies: 31
Director: Martin Scorsese, Number of Movies: 28


key points:
Directors with a high number of movies, such as Woody Allen, Alfred Hitchcok and etc., demonstrate their prolific nature and influence in the industry.
Trends in their work can reflect changes in audience preferences or technological advancements.


Code explanation:

%timeit: Measures the time taken to calculate the result.

links.top_directors(5): Returns a dictionary with the top 5 directors as keys and their respective movie counts as values.

For Loop: Prints each director and the number of movies they directed.

Purpose: Identifying prolific directors provides insights into their influence and contribution to the industry.



Performance
The method for identifying the top directors performed efficiently:

Execution Time: 6.91 ms ± 359 µs per loop (mean ± standard deviation over 7 runs with 100 loops each).

Analysis

Most Prolific Director:

Woody Allen leads with 46 movies, reflecting his long and highly productive career in filmmaking.

Legendary Filmmakers:

Alfred Hitchcock (37 movies) and Steven Spielberg (32 movies) underscore their significant contributions to the art of cinema, spanning multiple decades.

Diverse Styles:

Directors such as Clint Eastwood and Martin Scorsese showcase varied styles, from thought-provoking dramas to action-packed blockbusters, emphasizing their versatility and audience appeal.

Classic and Modern Influence:

The list includes directors from both classic (Hitchcock, Allen) and modern eras (Spielberg, Scorsese), highlighting enduring talent across different periods.

Conclusion

The data underscores the profound impact of these directors on the film industry. Their prolific outputs not only shaped genres but also created iconic works that resonate across generations. The variety in their styles and eras showcases the evolving landscape of cinema.



Budgets are a crucial factor in film production, reflecting the resources allocated to create cinematic experiences.

In [5]:
# Analyze the most expensive movies
%timeit expensive_movies = links.most_expensive(5)

# Display the movies
expensive_movies = links.most_expensive(5)
print("Most Expensive Movies:")
for movie, budget in expensive_movies.items():
    print(f"Movie: {movie}, Budget: {budget}")


37.3 ms ± 6.69 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Most Expensive Movies:
Movie: The Name of the Rose (1986), Budget: 30000000000.0
Movie: Hayot go'zal (1997), Budget: 15000000000.0
Movie: Gwoemul (2006), Budget: 12215500000.0
Movie: Cho'ri (2016), Budget: 10000000000.0
Movie: Pusan ekspressi (2016), Budget: 10000000000.0


Keypoints:

The most expensive movies, such as The name of the Rose, Hayot go'zal and etc., often represent high stakes in production, showcasing cutting-edge technology, large casts, or elaborate sets.
Their success or failure can significantly impact the industry's trajectory.


Code explanation:

%timeit: Evaluates the performance of links.most_expensive().

links.most_expensive(5): Returns a dictionary with the top 5 most expensive movies as keys and their budgets as values.

For Loop: Prints each movie and its budget.

Purpose: High-budget movies often signify ambitious projects, offering insights into industry investments and risks.

Performance

The method for identifying the most expensive movies performed as follows:

Execution Time: 37.3 ms ± 6.69 ms per loop (mean ± standard deviation over 7 runs with 10 loops each).

Analysis

Significant Budgets:

The Name of the Rose (1986) tops the list with an astonishing budget of 30 billion. This reflects a massive investment, possibly in sets, cast, or special effects.

Other movies such as Hayot go'zal (1997) and Gwoemul (2006) also had substantial budgets exceeding 10 billion.

Modern Films:

Two movies from 2016 (Cho'ri and Pusan ekspressi) highlight a trend in high-budget productions in recent years, potentially reflecting advancements in visual effects and global distribution models.

Diverse Origins:

The list includes movies from diverse countries and languages, showcasing the global nature of high-budget productions.

Conclusion

High-budget movies signify ambitious projects that aim to deliver exceptional visual and storytelling experiences. These investments reflect the evolving capabilities of the industry and its ability to produce large-scale, globally appealing content. The data also underscores how specific periods and regions have embraced high-budget filmmaking.